src/coreclr/ilasm/writer.cpp

```diff
@@ -51,10 +52,44 @@ HRESULT Assembler::InitMetaData()
     if(FAILED(hr = m_pEmitter->QueryInterface(IID_IMetaDataImport2, (void**)&m_pImporter)))
         goto exit;
 
+    if (m_fDeterministic)
+    {
+        //
+        // In deterministic mode, the MVID will need to be stabilized for the metadata scope that
+        // was created above, and the ChangeMvid service that makes this possible is only available
+        // on the IMDInternalEmit interface.
+        //
+        // When the CLSID_CorMetaDataDispenser instance above has activated against a current
+        // clr.dll (which is the only supported configuration for the determinism feature), it is
+        // guaranteed that "m_pEmitter" is implemented by the RegMeta object that was created
+        // during the DefineScope call above, and it is guaranteed that this same RegMeta object
+        // also implements the required IMDInternalEmit interface.
+        //
+        // Any failure is unexpected and catastrophic, so print a noisy message and return an
+        // error (which generally fails the entire ilasm operation) if any failure occurs.
+        //
+
+        hr = m_pEmitter->QueryInterface(IID_IMDInternalEmit, (void**)&m_pInternalEmitForDeterministicMvid);
+
+        if (FAILED(hr) || (m_pInternalEmitForDeterministicMvid == NULL))
+        {
+            fprintf(stderr, "Unexpected: Failed to query the required MVID determinism interface: %X\n",hr);
+            hr = E_FAIL;
+            goto exit;
+        }
+    }
+
     if (m_fGeneratePDB)
     {
         m_pPortablePdbWriter = new PortablePdbWriter();
         if (FAILED(hr = m_pPortablePdbWriter->Init(m_pDisp))) goto exit;
+
+        if (m_fDeterministic)
+        {
+            // Default values for determinism.
+            m_pPortablePdbWriter->SetGuid(GUID());
+            m_pPortablePdbWriter->SetTimestamp(0);
```

Comment:
Does this match what Roslyn does, or does Roslyn compute deterministic random stamp from the file hash?

I suspect that the time stamp and GUID are used by tools like symbol indexers. This may break them or make them to perform poorly (e.g. if they use the time stamp and GUID as a hash to compute set of candidates).

Reply:
> Does this match what Roslyn does, or does Roslyn compute deterministic random stamp from the file hash?

I'm not sure where to look in Roslyn's code for this, but Jared Parson's [blog post](https://blog.paranoidcoding.org/2016/04/05/deterministic-builds-in-roslyn.html) on deterministic builds mentions Roslyn uses the file hash for the GUID/timestamp too, for the reasons you mentioned. I'll switch this over.